metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+        }
+
+        if (!metadataVersion.isDelegationTokenSupported()) {
+            // DelegationTokens are not supported in this metadata version
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(UNSUPPORTED_VERSION.code()));
+        }
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
+
+        if (myTokenInformation.maxTimestamp() < now || myTokenInformation.expiryTimestamp() < now) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_EXPIRED.code()));
+        }
+
+        if (!allowedToRenew(myTokenInformation, context.principal())) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_OWNER_MISMATCH.code()));
+        }
+
+        if (requestData.expiryTimePeriodMs() < 0) { // expire immediately
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(requestData.expiryTimePeriodMs());
+            records.add(new ApiMessageAndVersion(new RemoveDelegationTokenRecord().
+                setTokenId(myTokenInformation.tokenId()), (short) 0));
+        } else {
+            long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(),
+                now + requestData.expiryTimePeriodMs());
+
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(expiryTimestamp);
+
+            DelegationTokenData newDelegationTokenData = new DelegationTokenData(myTokenInformation);
+            records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord()
+                .setExpirationTimestamp(expiryTimestamp), (short) 0));
+        }
+
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    // Periodic call to remove expired DelegationTokens
+    public List<ApiMessageAndVersion> sweepExpiredDelegationTokens() {
+        long now = time.milliseconds();
+        List<ApiMessageAndVersion> records = new ArrayList<ApiMessageAndVersion>();
+
+        for (TokenInformation oldTokenInformation: tokenCache.tokens()) {
+            if ((oldTokenInformation.maxTimestamp() < now) ||
+                (oldTokenInformation.expiryTimestamp() < now)) {
+                log.info("Delegation token expired for token: {} for owner: {}",
+                    oldTokenInformation.tokenId(), oldTokenInformation.ownerAsString());
+                records.add(new ApiMessageAndVersion(new RemoveDelegationTokenRecord().
+                    setTokenId(oldTokenInformation.tokenId()), (short) 0));
+            }
+        }
+        return records;
+    }
+
+    public void replay(DelegationTokenRecord record) {
+        log.info("Replayed DelegationTokenRecord for {}.", record.tokenId());
```

Comment:
Do we not have to do anything here?

Reply:
No, normally the replay here would add it to some hashtable to keep a cache of the state. We already have that in the tokenCache so adding another one here would just be wasting memory.